Network/Socket/ByteString/IO.hsc

```diff
@@ -248,3 +266,87 @@ withIOVec cs f =
         unsafeUseAsCStringLen s $ \(sPtr, sLen) ->
         poke ptr $ IOVec sPtr (fromIntegral sLen)
 #endif
+
+-- | Send data from the socket using sendmsg(2).
+sendMsg :: Socket       -- ^ Socket
+        -> SockAddr     -- ^ Destination address
+        -> [ByteString] -- ^ Data to be sent
+        -> [Cmsg]       -- ^ Control messages
+        -> [MsgFlag]    -- ^ Message flags
+        -> IO Int       -- ^ The length actually sent
+sendMsg _ _    []  _ _ = return 0
+sendMsg s addr bss cmsgs flags = do
+  sz <- withSockAddr addr $ \addrPtr addrSize ->
+    withIOVec bss $ \(iovsPtr, iovsLen) -> do
+      withCmsgs cmsgs $ \ctrlPtr ctrlLen -> do
+        let msgHdr = MsgHdr {
+                msgName    = addrPtr
+              , msgNameLen = fromIntegral addrSize
+              , msgIov     = iovsPtr
+              , msgIovLen  = fromIntegral iovsLen
+              , msgCtrl    = castPtr ctrlPtr
+              , msgCtrlLen = fromIntegral ctrlLen
+              , msgFlags   = 0
+              }
+            cflags = packBits msgFlagMapping flags
+        withFdSocket s $ \fd ->
+          with msgHdr $ \msgHdrPtr ->
+            throwSocketErrorWaitWrite s "Network.Socket.ByteString.sendMsg" $
+              c_sendmsg fd msgHdrPtr cflags
+  return $ fromIntegral sz
+
+-- | Receive data from the socket using recvmsg(2).
+--   The receive buffers are created according to the second argument.
+--   If the length of received data is less than the total of
+--   the second argument, the buffers are truncated properly.
+--   So, only the received data can be seen.
+recvMsg :: Socket    -- ^ Socket
+        -> [Int]     -- ^ A list of length of data to be received
+                     --   If the total length is not large enough,
+                     --   'MSG_TRUNC' is returned
```

Comment:
Is there any point in using scatter/gather iovecs here?  This may make sense in the C API, but it is unclear why a Haskell user would want to receive the message as an array of fragments...

Reply:
I was interested in having this feature when I worked on a bittorrent client. It's certainly low level, but can be useful.

Reply:
Some more substance is needed here, what __specifically__ makes scatter/gather useful here?  ByteStrings support efficient slices in Haskell, what benefit is there in returning multiple pre-sliced bytestrings?

Reply:
I have no preference on either `Int` or `[Int]`. :-)

Reply:
I can see how perhaps in trying to minimize copying one might use scatter/gather on the write side, when say sending some data to which one needs to prefix a header.  On the read side, it just does not seem to be useful, and I think makes the interface needlessly complex.  Perhaps I'm missing something, but I think that a single read buffer is quite sufficient and simpler.

Reply:
The advantage is that you can do a single system call and have the kernel write to the exact locations of your buffers. If you recv all of your data into one buffer, you tend to have to copy it around later, in your program. The kernel can do it all in one step.

Reply:
> The advantage is that you can do a single system call and have the kernel write to the exact locations of your buffers. If you recv all of your data into one buffer, you tend to have to copy it around later, in your program. The kernel can do it all in one step.

That's true in C, but really can't any use for that in Haskell. As I already mentioned, Haskell has efficient `ByteString` slices, so logically you can divide up a ByteString into as many pieces as you want without copying the data. There's probably more overhead in the logic for building all the arrays, then in any benefit from creating pre-partitioned byte strings.

Again, you're speaking vague generalities, what specific use-case benefits from this?

Reply:
mmap()-ing files and recv()-ing iovecs from that file's address range

Reply:
> mmap()-ing files and recv()-ing iovecs from that file's address range

You could do that in C, but __this__ interface can't do that, it returns an array of freshly allocated ByteStrings at addresses you have no control over.  All that's happening here is that the ByteStrings are artificially fragmented into a few chunks, for no obvious gain. The function __does not__ mutate the content of a given `iovec`.

Reply:
OK, got it. It's not very useful having `[Int]` here, then.

Reply:
> OK, got it. It's not very useful having `[Int]` here, then.

Yes, thanks, that's the reason I persisted in questioning the choice. This should always pass a single `iobuf` of the requested length on the read side. On the write side scatter/gather can make some sense, if there's enough interest in sending multiple fragments without copying. We could even provide two functions one that takes a single message (runs less code), and a second that builds `iovecs` around the buffers of multiple `ByteStrings`. IIRC `ByteString` contents are in pinned memory, and the `Internal` ByteString module exposes their pointers, so the writes could be directly from the ByteString buffers.

Reply:
Done.